### PR TITLE
Kernel

### PR DESCRIPTION
**Pull request by gvarao**
- Added Kernel playbooks

### CODE REVIEW

**Comment by gvarao** [comment]
jenkins please retry a build

File: juniper_official/Kernel/krt-statistics-kpis.playbook [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ *     <rpc>
+ *         <get-async-cor-stats>
+ *         </get-async-cor-stats>
+ *     </rpc>
+ *     <cli>
+ *         <banner></banner>
+ *     </cli>
+ * </rpc-reply>
+ * 
+ * labroot@jtac-mx960> show krt state | display xml
```
Comment (ConstrainedRandomVar) [inline juniper_official/Kernel/krt-statistics-kpis.playbook]: Nice.. I like this view of the device-output, though I can't decide if this should be here in the playbook file or in the rule file. 
When we introduce the playbook-abstraction feature, the goal is to allow a playbook to have #ifdef compilations/logic built in for rules depending on the meta-data properties. So, for instance, I can have a single interface-KPIs playbook pull in OC rules for MX >= 17.1; iAgent rules for MX< 17.1 && all SRX, etc..

File: juniper_official/Kernel/check-krt-ack-timeouts.rule [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                    value "{{error-threshold}}";
+                }
+                type integer;
+                description "KRT route and next-hop timeout count increase threshold";
+            }
+            trigger tr-timeouts {
+                frequency 60s;
+                term route-timeouts {
+                    when {
+                        increasing-at-least-by-value "$krt-rt-acknowledgement-timeouts" {
```
[the remark this answers is not in the document]
Comment (gvarao) [reply juniper_official/Kernel/check-krt-ack-timeouts.rule]: Addressed.

File: juniper_official/Kernel/krt-statistics-kpis.playbook (hunk 1 above)
Comment (gvarao) [reply juniper_official/Kernel/krt-statistics-kpis.playbook]: Agree with you

File: juniper_official/Kernel/check-krt-ack-timeouts.rule (hunk 2 above)
[the remark this answers is not in the document]
Comment (gvarao) [reply juniper_official/Kernel/check-krt-ack-timeouts.rule]: Added and committed

**Comment by gvarao** [comment]
jenkins please retry a build

**Comment by gvarao** [comment]
jenkins please retry a build

File: juniper_official/Kernel/check-krt-ack-timeouts.rule [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                term nexthop-timeouts {
+                    when {
+                        increasing-at-least-by-value "$krt-nh-acknowledgement-timeouts" {
+                            value "$threshold";
+                        }
+                    }
+                    then {
+                        status {
+                            color red;
+                            message " krt nexthop acknowledgement-timeouts are increasing.Value =($krt-nh-acknowledgement-timeouts) ";
```
Comment (ConstrainedRandomVar) [inline juniper_official/Kernel/check-krt-ack-timeouts.rule]: Issue with whitespace.

File: juniper_official/Kernel/check-pfeman-conn-drops.rule [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * with Master Routing Engine Junos Kernel.
+ *
+ * This rule monitors and ungraceful connection drops in these pfeman connections.
+ */
+healthbot {
+    topic kernel.peer-infra {
+        synopsis "Monitor peer-infra activities on the system";
+        rule check-pfeman-conn-drops {
+            sensor kernel-peer-infra-sensor {
+                synopsis "peer-infra(pfeman) connection drops open-config sensor";
```
Comment (ConstrainedRandomVar) [inline juniper_official/Kernel/check-pfeman-conn-drops.rule]: Whitespace.

File: juniper_official/Kernel/check-stuck-ifstate-clients.rule [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                description "monitor and report any stuck ifstate clients on the network device";
+                frequency 90s;
+                term is-ifs-client-stuck {
+                    when {
+                        not-equal-to "$stuck-clients" 0;
+                    }
+                    then {
+                        status {
+                            color red;
+                            message "Found stuck daemon/fpc. For details run: show system alarms. Restart stuck clients for recovery.";
```
Comment (ConstrainedRandomVar) [inline juniper_official/Kernel/check-stuck-ifstate-clients.rule]: Would it make sense to collect the output of "show system alarm" through another rule and just reference it here in the message - stating that here is the daemon/fpc that is stuck"?

Comment (gvarao) [reply juniper_official/Kernel/check-stuck-ifstate-clients.rule]: Checked with prateek. It is difficult to predict fpc or daemon and also not possible to get information from other rules as it is based on event.

File: juniper_official/Kernel/check-pfeman-conn-drops.rule (hunk 4 above)
Comment (gvarao) [reply juniper_official/Kernel/check-pfeman-conn-drops.rule]: Addressed

File: juniper_official/Kernel/check-krt-ack-timeouts.rule [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                        }
+                    }
+                }
+                term no-timeouts {
+                    then {
+                        status {
+                            color green;
+                            message "Route ack and Nexthop ack timeouts are not increasing. ";
```
[the remark this answers is not in the document]
Comment (gvarao) [reply juniper_official/Kernel/check-krt-ack-timeouts.rule]: Addressed

File: juniper_official/Kernel/check-krt-ack-timeouts.rule (hunk 3 above)
Comment (gvarao) [reply juniper_official/Kernel/check-krt-ack-timeouts.rule]: Addressed

File: juniper_official/Kernel/check-client-limit-reached.rule [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            field client-limit-reached {
+                sensor kernel-ifstate-sensor {
+                    path /junos/kernel-ifstate/client-limit-reached;
+                }
+                type integer;
+                description "Client limit reached";
+                }
+            trigger report-ifstate-client-limit-reached {
+                synopsis "report if any application has opened more than allowed number of ifstate based routing sockets";
+                description "raise a RED alarm if a application has opened more than allowed number of ifstate based routing sockets";
```
[the remark this answers is not in the document]
Comment (gvarao) [reply juniper_official/Kernel/check-client-limit-reached.rule]: Addressed

File: juniper_official/Kernel/check-client-limit-reached.rule [hunk 8]
```diff
@@ -0,0 +1,54 @@
+/*
+ * This rule checks if any application has opened more than the allowed number
+ * of Ifstate based routing sockets.
+ */
+healthbot {
+    topic kernel.ifstate {
+        rule check-client-limit-reached {
+            sensor kernel-ifstate-sensor {
+                synopsis "number of ifstate based routing sockets for an application";
```
[the remark this answers is not in the document]
Comment (gvarao) [reply juniper_official/Kernel/check-client-limit-reached.rule]: Addressed

File: juniper_official/Kernel/check-stuck-ifstate-clients.rule (hunk 5 above)
Comment (gvarao) [reply juniper_official/Kernel/check-stuck-ifstate-clients.rule]: Checked with prateek. It is difficult to predict fpc or daemon and also not possible to get information from other rules as it is based on event.

**Pull request merged**